crates/pkvdb/src/lib.rs

```diff
@@ -0,0 +1,451 @@
+    }
+
+    pub fn purge(&mut self) {
+        // TODO:george this approach is not sound for performance
+        if let Ok(mut db_mut) = self.leveldb.lock() {
+            let mut writebatch = WriteBatch::new();
+            if let Ok(mut iter) = DB::new_iter(db_mut.borrow_mut()) {
+                while let Some((key, val)) = iter.next() {
+                    // FIXME: if we should keep the negative reference count
+                    if i32::from_be_bytes(val[0..4].try_into().unwrap()) == 0 {
+                        writebatch.delete(&key);
+                    }
+                }
+                let _ = DB::write(db_mut.borrow_mut(), writebatch, true);
+            }
+        }
+    }
+
+    // only for test
+    pub fn clear(&mut self) {
+        if let Ok(mut db_mut) = self.leveldb.lock() {
+            let mut writebatch = WriteBatch::new();
+            if let Ok(mut iter) = DB::new_iter(db_mut.borrow_mut()) {
+                while let Some((key, _)) = iter.next() {
+                    writebatch.delete(&key);
+                }
+                let _ = DB::write(db_mut.borrow_mut(), writebatch, true);
+            }
+        }
+    }
+
+    pub fn raw(&self, key: &H::Out, _prefix: Prefix) -> Option<(T, i32)> {
+        if key == &self.hashed_null_node {
+            return Some((self.null_node_data.clone(), 1));
+        }
+        let mut db_mut = self.leveldb.lock().unwrap();
+        match DB::get(db_mut.borrow_mut(), key.as_ref()).map(Inner::<T>::from) {
+            Some(Inner(value, rc)) => Some((value, rc)),
+            _ => None,
+        }
+    }
+}
+
+impl<H> Kvdb<H, sp_trie::DBValue>
+where
+    H: Hasher,
+{
+    pub fn consolidate(&mut self, mut transaction: Transaction<H>) {
+        if let Ok(mut db_mut) = self.leveldb.lock() {
+            let mut writebatch = WriteBatch::new();
+
+            for (key, (value, rc)) in transaction.drain() {
+                match DB::get(db_mut.borrow_mut(), key.as_ref())
+                    .map(Inner::<sp_trie::DBValue>::from)
+                {
+                    Some(mut inner) => {
+                        inner.1 += rc;
+                        writebatch.put(key.as_ref(), Into::<Vec<u8>>::into(inner).as_ref());
+                    }
+                    None => {
+                        let inner = Inner::<sp_trie::DBValue>(value, rc);
+                        writebatch.put(key.as_ref(), Into::<Vec<u8>>::into(inner).as_ref());
+                    }
+                }
+            }
+            if writebatch.count() > 0 {
+                let _ = DB::write(db_mut.borrow_mut(), writebatch, true);
+            }
+        }
+    }
+}
+
+// inner kv represention
+// first 4 bytes is the reference count and left data vector is the data self
+struct Inner<T>(T, i32);
+
+impl<T> From<Vec<u8>> for Inner<T>
+where
+    T: for<'a> From<&'a [u8]>,
+{
+    fn from(v: Vec<u8>) -> Self {
+        let (rc, val) = v.split_at(4);
+        let value = T::from(val);
```

Comment:
Given the inner T is usually a `sp_trie::DBValue` (alias of `Vec<u8>`), this looks not very performant, becase it a would allocate at least twice on each get.
Though, I haven't an idea yet on what we can do to optimize.